### DISCUSSION TITLE
AI deck unable to flash after running my own code

### DISCUSSION THREAD

**Discussion by pladosz** [bitcraze/discussions#542] [Q&A] [answered]
Hi,

I seem to have put myself in a bit of predicament. I was testing methods of how to send velocity commands over CPX to the crazyflie from AI deck (following earlier discussion #519 ). To do so,  I wrote an initial testing code on the crazyflie (see the testing code below, please DO NOT run it, it is here for reference only). Since I flashed it, I am unable to flash anything to the AI deck over crazyradio, it gets stuck at ‘Reset to bootloader mode...’ . I seem to trigger a lot of cpx errors about invalid targets which are continuously printed in crazyflie console. I suspect what is happening is crazyradio is overloaded with error messages and thus is unable to do anything else. I have tried flashing wi-fi streaming example to AI-deck over JTAG, which completes successfully, but that does not seem to erase my testing code. I can also flash firmware to GAP 8 using JTAG (following initial setup guide here), but that also does not erase the problematic code. (Just for full information, I have since fixed the issue in my code by placing timers to reduce the frequency of messages). Would flashing my improved code over JTAG fix the issue (I am a bit reluctant to try in case it makes things worse)? If not, I wonder if there is any solution to this problem? (for example erasing AI deck’s memory through JTAG or even a factory reset). Thank you for all the help!

```#include "pmsis.h"

#include "bsp/bsp.h"
#include "bsp/camera/himax.h"
#include "bsp/buffer.h"
#include "stdio.h"

#include "cpx.h"
#include "wifi.h"

static pi_task_t task1;

static struct pi_device camera;
static pi_buffer_t buffer;

static EventGroupHandle_t evGroup;
#define CAPTURE_DONE_BIT (1 << 0)

// Performance menasuring variables
static uint32_t start = 0;
static uint32_t transferTime = 0;



static void capture_done_cb(void *arg)
{
  xEventGroupSetBits(evGroup, CAPTURE_DONE_BIT);
}

typedef struct
{
  uint16_t Type;
  float vx;
  float vy;
  float vz;
  float yawrate;
} __attribute__((packed)) vel_message_t;


static CPXPacket_t txp;

void createVelMsgPacket(CPXPacket_t * packet) {
  vel_message_t *velMsg = (vel_message_t *) packet->data;
  velMsg->Type = 1;
  velMsg->vx = 0.1;
  velMsg->vy = 0.1;
  velMsg->vz = 0.1;
  velMsg->yawrate = 0;
  packet->dataLength = sizeof(vel_message_t);
}

void sendBufferViaCPX(CPXPacket_t * packet, uint8_t * buffer, uint32_t bufferSize) {
  uint32_t offset = 0;
  uint32_t size = 0;
  do {
    size = sizeof(packet->data);
    if (offset + size > bufferSize)
    {
      size = bufferSize - offset;
    }
    memcpy(packet->data, &buffer[offset], sizeof(packet->data));
    packet->dataLength = size;
    cpxSendPacketBlocking(packet);
    offset += size;
  } while (size == sizeof(packet->data));
}




void velocity_task(void *parameters) {
  //initalize CPX route
  cpxInitRoute(CPX_T_GAP8, CPX_T_STM32, CPX_F_CRTP, &txp.route);
  while (1) {
    // send velocity image
    cpxPrintToConsole(LOG_TO_CRTP, "attempting to send velocity !\n");
    createVelMsgPacket(&txp);
    cpxSendPacketBlocking(&txp);
  }

}

void start_example(void)
{
    struct pi_uart_conf conf;
  struct pi_device device;
  pi_uart_conf_init(&conf);
  conf.baudrate_bps = 115200;

  pi_open_from_conf(&device, &conf);
  if (pi_uart_open(&device))
  {
    printf("[UART] open failed !\n");
    pmsis_exit(-1);
  }

  cpxInit();
  cpxEnableFunction(CPX_F_CRTP);

  evGroup = xEventGroupCreate();

  BaseType_t xTask;

  xTask = xTaskCreate(velocity_task, "vel_sending_task", configMINIMAL_STACK_SIZE * 2,
                      NULL, tskIDLE_PRIORITY + 1, NULL);
  if (xTask != pdPASS)
  {
    cpxPrintToConsole(LOG_TO_CRTP, "velocity sending task didn't start !\n");
    pmsis_exit(-1);
  }
  while (1)
  {
    pi_yield();
  }
}


int main(void)
{
  pi_bsp_init();

  // Increase the FC freq to 250 MHz
  pi_freq_set(PI_FREQ_DOMAIN_FC, 250000000);
  pi_pmu_voltage_set(PI_PMU_DOMAIN_FC, 1200);

  return pmsis_kickoff((void *)start_example);
}```

**Comment by hmllr** [comment]
Hi :)
What do you mean exactly by "it does not erase the code"? You cannot have two programs flashed at the same time (at least without any major changes to them), so this doesn't seem to be possible. Maybe you are only running the example code and not flashing it (you always execute from RAM in gap8, so the code that runs doesn't necessarily have to be written to flash). 
Anyway, I would try to re-flash the bootloader (over JTAG): https://github.com/bitcraze/aideck-gap8-bootloader and then to update OTA again.
Hope this helps and let us know if it does,
Hanna

**Reply by pladosz** [reply]
Hi, I haven't had time to retest those things. I will update as soon as I manage

**Comment by bhabas** [comment, marked as answer]
We were able to work out a solution to a very similar problem here (#637) if you're still interested @pladosz

**Reply by pladosz** [reply]
Thanks, we ended up using JTAG as a workaround but your solution is very helpful!
